Data/Text/ICU/Translit/IO.hs

```diff
@@ -1,54 +1,70 @@
 module Data.Text.ICU.Translit.IO
-    (Transliterator,
-     transliterator,
-     transliterate) where
+  ( Transliterator
+  , transliterator
+  , transliterate
+  )
+where
 
-import Foreign
-import Data.Text
-import Data.Text.Foreign
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as BS
+import Data.Text (Text)
+import Data.Text.Encoding qualified as T
 import Data.Text.ICU.Translit.ICUHelper
+  ( UChar
+  , UErrorCode
+  , handleError
+  , handleFilledOverflowError
+  )
+import Foreign
 
 data UTransliterator
 
-foreign import ccall "trans.h __hs_translit_open_trans" openTrans
+foreign import ccall "trans.h __hs_translit_open_trans"
+  openTrans
     :: Ptr UChar -> Int -> Ptr UErrorCode -> IO (Ptr UTransliterator)
-foreign import ccall "trans.h &__hs_translit_close_trans" closeTrans
-    :: FunPtr (Ptr UTransliterator -> IO ())
-foreign import ccall "trans.h __hs_translit_do_trans" doTrans
-    :: Ptr UTransliterator -> Ptr UChar -> Int32 -> Int32
-    -> Ptr UErrorCode -> IO Int32
-
 
+foreign import ccall "trans.h &__hs_translit_close_trans"
+  closeTrans
+    :: FunPtr (Ptr UTransliterator -> IO ())
 
-data Transliterator = Transliterator {
-      transPtr :: ForeignPtr UTransliterator,
-      transSpec :: Text
-    }
+foreign import ccall "trans.h __hs_translit_do_trans"
+  doTrans
+    :: Ptr UTransliterator
+    -> Ptr UChar
+    -> Int32
+    -> Int32
+    -> Ptr UErrorCode
+    -> IO Int32
 
+data Transliterator = Transliterator
+  { transPtr :: ForeignPtr UTransliterator
+  , transSpec :: Text
+  }
 
 instance Show Transliterator where
-    show tr = "Transliterator " ++ show (transSpec tr)
-
-
+  show tr = "Transliterator " ++ show (transSpec tr)
 
+-- we just assume little endian
 transliterator :: Text -> IO Transliterator
-transliterator spec =
-    useAsPtr spec $ \ptr len -> do
-           q <- handleError $ openTrans ptr (fromIntegral len)
-           ref <- newForeignPtr closeTrans q
-           touchForeignPtr ref
-           return $ Transliterator ref spec
-
+transliterator spec = do
+  let specStr :: ByteString = T.encodeUtf16LE spec
+  BS.useAsCStringLen specStr $ \((castPtr @_ @Word16) -> ptr, (`div` 2) -> len) -> do
+    q <- handleError $ openTrans ptr (fromIntegral len)
+    ref <- newForeignPtr closeTrans q
+    touchForeignPtr ref
```

Comment:
Like we discussed, I'm not sure this does anything. Does something break if you remove it?

Reply:
(I mean the call `touchForeignPtr`, not the whole block :))